server/api/student/student.controller.js

```diff
@@ -84,6 +87,123 @@ exports.outreachCounts = function(req, res) {
   });
 };
 
+function createNameAndSchoolPromise(results) {
+  return new Promise(function(resolve, reject) {
+    var promises = _.map(results, function(result) {
+      return new Promise(function(resolve, reject) {
+        function getTotals() {
+          var records = this.records;
+          var typeArr = _.map(records, 
+              function(rec) { return rec.type 
+            });
+          return _.countBy(typeArr, _.identity);
+        }
+        School.findById(result.school, 
+          function(err, school) {
+            if (err) reject(err);
+            result.school = school;
+          })
+          .then(function() {
+            Student.findById(result.student, 
+              function(err, student) {
+              if (err) reject(err);
+              result.student = student;
+            })
+          .then(function() {
+            result.totals = getTotals.call(result);
+            resolve(result);
+          });
+        });
+      });
+    });
+    Promise.all(promises).then(
+      function(val) { 
+        resolve(val);
+    });
+  });
+}
+
```

Comment:
This whole function seems a little excessive in terms of promises, but I couldn't think of a more elegant way to run all these queries asynchronously. I believe it also is very poor in terms of efficiency and causes lag when loading in the school reports page. Tried to refactor down some of the promises by using the built in `.then()` statement.